src/events.ts

```diff
@@ -0,0 +1,81 @@
+import { MraidState } from "./state";
+
+export enum MraidEvent {
+  Ready = "ready",
+  Error = "error",
+  StateChange = "stateChange",
+  ViewableChange = "viewableChange",
+}
+
+export type EventListener<T1, T2> = (t1: T1, t2: T2) => void;
+export type ErrorEventListener = EventListener<string, string | null>;
+export type StateChangeEventListener = EventListener<MraidState, void>;
+export type ViewableChangeEventListener = EventListener<boolean, void>;
+export type ReadyEventListener = EventListener<void, void>;
+
+export type MraidEventListener =
+  | ErrorEventListener
+  | StateChangeEventListener
+  | ViewableChangeEventListener
+  | ReadyEventListener
+  | null
```

Comment:
I'm not sure to understand why you can have null or undefined values as listener. By accepting those values, then, when using the listener, you need to prepare yourself against null/undefined values (while you're not having them in the `eventListeners` sets).

I would more expect to not handle those, and to simply make the removeEventListener method accept either a listener or null/undefined.

Reply:
Changed to what you've suggested + added guards when passing values from js